src/server/pps/server/api_server.go

```diff
@@ -113,87 +116,85 @@ type apiServer struct {
-				return
+		if err := func() error {
+			set := false
+			if input.Atom != nil {
+				set = true
+				switch {
+				case len(input.Atom.Name) == 0:
+					return fmt.Errorf("input must specify a name")
+				case input.Atom.Name == "out":
+					return fmt.Errorf("input cannot be named \"out\", as pachyderm " +
+						"already creates /pfs/out to collect job output")
+				case input.Atom.Repo == "":
+					return fmt.Errorf("input must specify a repo")
+				case input.Atom.Branch == "" && !job:
+					return fmt.Errorf("input must specify a branch")
+				case input.Atom.Commit == "" && job:
+					return fmt.Errorf("input must specify a commit")
+				case len(input.Atom.Glob) == 0:
+					return fmt.Errorf("input must specify a glob")
+				}
+				if _, ok := names[input.Atom.Name]; ok {
+					return fmt.Errorf("conflicting input names: %s", input.Atom.Name)
+				}
+				names[input.Atom.Name] = true
+				if repoBranch[input.Atom.Repo] != "" && repoBranch[input.Atom.Repo] != input.Atom.Branch {
+					return fmt.Errorf("cannot use the same repo in multiple inputs with different branches")
+				}
+				repoBranch[input.Atom.Repo] = input.Atom.Branch
+				if job {
+					// for jobs we check that the input commit exists
+					if _, err := pachClient.InspectCommit(input.Atom.Repo, input.Atom.Commit); err != nil {
+						return err
+					}
+				} else {
+					// for pipelines we only check that the repo exists
+					if _, err = pachClient.InspectRepo(input.Atom.Repo); err != nil {
+						return err
+					}
+				}
 			}
-			names[input.Atom.Name] = true
-			if repoBranch[input.Atom.Repo] != "" && repoBranch[input.Atom.Repo] != input.Atom.Branch {
-				result = fmt.Errorf("cannot use the same repo in multiple inputs with different branches")
-				return
+			if input.Cross != nil {
+				if set {
+					return fmt.Errorf("multiple input types set")
+				}
+				set = true
 			}
-			repoBranch[input.Atom.Repo] = input.Atom.Branch
-			pfsClient, err := a.getPFSClient()
-			if err != nil {
-				result = err
-				return
+			if input.Union != nil {
+				if set {
+					return fmt.Errorf("multiple input types set")
+				}
+				set = true
 			}
-			if job {
-				// for jobs we check that the input commit exists
-				_, err = pfsClient.InspectCommit(ctx, &pfs.InspectCommitRequest{
-					Commit: client.NewCommit(input.Atom.Repo, input.Atom.Commit),
-				})
-				if err != nil {
-					result = err
-					return
+			if input.Cron != nil {
+				if set {
+					return fmt.Errorf("multiple input types set")
 				}
-			} else {
-				// for pipelines we only check that the repo exists
-				_, err = pfsClient.InspectRepo(ctx, &pfs.InspectRepoRequest{
-					Repo: client.NewRepo(input.Atom.Repo),
-				})
-				if err != nil {
-					result = err
-					return
+				set = true
+				if _, err := cron.Parse(input.Cron.Spec); err != nil {
+					return err
+				}
+				if err := pachClient.CreateRepo(input.Cron.Repo); err != nil && !strings.Contains(err.Error(), "already exists") {
```

Comment:
I'm cool with that, I was thinking that it has the disadvantage that you have to do `Visit` twice and you need both a call to `InpsectRepo` in validate and `CreateRepo` in `CreatePipeline`. But neither of these are really that big of a deal.